ms2deepscore/train_new_model/inchikey_pair_selection.py

```diff
@@ -120,44 +70,106 @@ def select_compound_pairs_wrapper(
+        pairs = []
+        for i, inchikey_index_1 in enumerate(inchikey_indexes_1):
+            inchikey_1 = inchikeys[inchikey_index_1]
+            inchikey_2 = inchikeys[selected_pairs_per_bin_matrix[bin_idx, inchikey_index_1, pair_sample_position[i]]]
+            score = scores_per_bin[bin_idx, inchikey_index_1, pair_sample_position[i]]
+            # Check that the reversed pair (which is the same) is not already in the set
+            if (inchikey_2, inchikey_1, score) not in pairs:
+                pairs.append((inchikey_1, inchikey_2, score))
+        selected_pairs_per_bin.append(pairs)
+    return selected_pairs_per_bin
+
+
+def balanced_selection_of_pairs_per_bin(list_of_pairs_per_bin,
+                                        unique_inchikeys):
+    inchikey_count = {inchikey: 0 for inchikey in unique_inchikeys}
+    sorted_bin_indices_on_amount_of_pairs = sorted(range(len(list_of_pairs_per_bin)),
+                                                   key=lambda i: len(list_of_pairs_per_bin[i]))
+    lowest_number_of_pairs = min(len(pairs) for pairs in list_of_pairs_per_bin)
+    selected_pairs_per_bin = []
+    for bin_index in tqdm(sorted_bin_indices_on_amount_of_pairs):
+        selected_pairs_per_bin.append(
+            select_pairs(list_of_pairs_per_bin[bin_index], inchikey_count, lowest_number_of_pairs))
+    return selected_pairs_per_bin
+
+
+def select_pairs(list_of_available_pairs, inchikey_counts, required_number_of_pairs):
+    selected_pairs = []
+    for _ in range(required_number_of_pairs):
+        # Select only the inchikeys that still have a pair available for this bin.
+        available_inchikey_indexes = get_available_inchikey_indexes(list_of_available_pairs)
+        # get lowest available inchikeys
+        inchikey_with_lowest_count = get_available_inchikeys_with_lowest_count(available_inchikey_indexes,
+                                                                               inchikey_counts)
+        # actually select pairs (instead of single inchikeys)
+        available_pairs_for_least_frequent_inchikey = select_available_pairs(list_of_available_pairs, inchikey_with_lowest_count)
+        idx_2_counts = [inchikey_counts[idx_2] for _, idx_2, _ in available_pairs_for_least_frequent_inchikey]
+        index_of_least_frequent_second_index = idx_2_counts.index(min(idx_2_counts))
+        idx_1, idx_2, score = available_pairs_for_least_frequent_inchikey[index_of_least_frequent_second_index]
+        # Remove the selected pair and a potential reversed pair if available
+        if (idx_1, idx_2, score) in list_of_available_pairs:
+            list_of_available_pairs.remove((idx_1, idx_2, score))
+        if (idx_2, idx_1, score) in list_of_available_pairs:
+            list_of_available_pairs.remove((idx_2, idx_1, score))
+
+        selected_pairs.append((idx_1, idx_2, score))
+        inchikey_counts[idx_1] += 1
+        inchikey_counts[idx_2] += 1
+    return selected_pairs
```

Comment:
`inchikey_counts` is given as a parameter and updated in this function --> should also be returned

Reply:
Ahh. Sorry, I see now that this is only needed with within the for-loop. I would then better make this clearer in the code (I'll add a suggestion).

Reply:
No you are correct! I try to achieve balanced inchikey counts over the different bins. The code does this currently, but it is always confusing and dangerous to update dictionaries within a function without returning. So I fully agree this should be changed.